File: core/src/test/java/chaij/IntExpectationTests.java
package chaij;


import org.junit.Rule;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.rules.ExpectedException;
import org.junit.runner.RunWith;

import static chaij.ChaiJ.expect;

@RunWith(Enclosed.class)
public class IntExpectationTests {
	
	@SuppressWarnings("UnnecessaryExplicitNumericCast")
	public static class NormalTests {
		
		@Test
		public void testEqual() {
			
			expect(42).equal(42);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongEqual() {
			
			expect(42).equal(43);
		}
		
		
		@Test
		public void testAbove() {
			
			expect(42).above(41);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongAbove() {
			
			expect(42).above(42);
		}
		
		
		@Test
		public void testLeast() {
			
			expect(42).least(42);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongLeast() {
			
			expect(42).least(43);
		}
		
		
		@Test
		public void testBelow() {
			
			expect(42).below(43);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongBelow() {
			
			expect(42).below(42);
		}
		
		
		@Test
		public void testMost() {
			
			expect(42).most(42);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongMost() {
			
			expect(42).most(41);
		}
		
		
		@Test
		public void testWithinExact() {
			
			expect(42).within(42, 42);
		}
		
		
		@Test
		public void testWithinNormal() {
			
			expect(42).within(41, 43);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongWithinSmaller() {
			
			expect(42).within(41, 41);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongWithinBigger() {
			
			expect(42).within(43, 43);
		}
		
		
		@Test
		public void testMatch() {
			
			expect(42).match(i -> i == 42);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongMatch() {
			
			expect(42).match(i -> i == 43);
		}
		
		
		@Test
		public void testSatisfy() {
			
			expect(42).satisfy(i -> i == 42);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongSatisfy() {
			
			expect(42).satisfy(i -> i == 43);
		}
		
		
		@Test
		public void testCloseToDistanceZero() {
			
			expect(42).closeTo(42, 0);
		}
		
		
		@Test
		public void testCloseToDistanceOne() {
			
			expect(42).closeTo(41, 1).closeTo(43, 1);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongCloseTo() {
			
			expect(42).closeTo(41, 0);
		}
		
		
		@Test
		public void testOneOf() {
			
			expect(42).oneOf(41, 42, 43);
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongOneOf() {
			
			expect(42).oneOf(41, 43);
		}
		
		
		@Test
		public void testValidByte() {
			
			expect((int) Byte.MAX_VALUE).validByte();
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongValidByte() {
			
			expect((int) Byte.MAX_VALUE + 1).validByte();
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongNegativeValidByte() {
			
			expect((int) Byte.MIN_VALUE - 1).validByte();
		}
		
		
		@Test
		public void testValidShort() {
			
			expect((int) Short.MAX_VALUE).validShort();
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongValidShort() {
			
			expect((int) Short.MAX_VALUE + 1).validShort();
		}
		
		
		@Test(expected = UnmetExpectationException.class)
		public void testWrongNegativeValidShort() {
			
			expect((int) Short.MIN_VALUE - 1).validShort();
		}
	}
	
	
	public static class MessageTests {
		
		@Rule
		public final ExpectedException e = ExpectedException.none();
		
		
		@Test
		public void testStandard() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Special: Expected 42 to equal 43.");
			expect(42, "Special").equal(43);
		}
		
		
		@Test
		public void testEqual() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to equal 43.");
			expect(42).equal(43);
		}
		
		
		@Test
		public void testEqualWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to equal 43.");
			expect(42).equal(43, "Custom");
		}
		
		
		@Test
		public void testAbove() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to be above 43.");
			expect(42).above(43);
		}
		
		
		@Test
		public void testAboveWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to be above 43.");
			expect(42).above(43, "Custom");
		}
		
		
		@Test
		public void testLeast() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to be at least 43.");
			expect(42).least(43);
		}
		
		
		@Test
		public void testLeastWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to be at least 43.");
			expect(42).least(43, "Custom");
		}
		
		
		@Test
		public void testBelow() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to be below 41.");
			expect(42).below(41);
		}
		
		
		@Test
		public void testBelowWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to be below 41.");
			expect(42).below(41, "Custom");
		}
		
		
		@Test
		public void testMost() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to be at most 41.");
			expect(42).most(41);
		}
		
		
		@Test
		public void testMostWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to be at most 41.");
			expect(42).most(41, "Custom");
		}
		
		
		@Test
		public void testWithin() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to be within 43 and 43.");
			expect(42).within(43, 43);
		}
		
		
		@Test
		public void testWithinWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to be within 43 and 43.");
			expect(42).within(43, 43, "Custom");
		}
		
		
		@Test
		public void testMatch() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to match a custom predicate.");
			expect(42).match(i -> false);
		}
		
		
		@Test
		public void testMatchWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to match a custom predicate.");
			expect(42).match(i -> false, "Custom");
		}
		
		
		@Test
		public void testSatisfy() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to satisfy a custom predicate.");
			expect(42).satisfy(i -> false);
		}
		
		
		@Test
		public void testSatisfyWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to satisfy a custom predicate.");
			expect(42).satisfy(i -> false, "Custom");
		}
		
		
		@Test
		public void testCloseTo() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to be close to 43 with a delta of 0.");
			expect(42).closeTo(43, 0);
		}
		
		
		@Test
		public void testCloseToWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to be close to 43 with a delta of 0.");
			expect(42).closeTo(43, 0, "Custom");
		}
		
		
		@Test
		public void testOneOf() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 42 to be one of [41, 43].");
			expect(42).oneOf(41, 43);
		}
		
		
		@Test
		public void testOneOfWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 42 to be one of [41, 43].");
			expect(42).oneOf("Custom", 41, 43);
		}
		
		
		@Test
		public void testValidByte() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 128 to be a valid byte value.");
			expect(Byte.MAX_VALUE + 1).validByte();
		}
		
		
		@Test
		public void testValidByteWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 128 to be a valid byte value.");
			expect(Byte.MAX_VALUE + 1).validByte("Custom");
		}
		
		
		@Test
		public void testValidShort() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Expected 32768 to be a valid short value.");
			expect(Short.MAX_VALUE + 1).validShort();
		}
		
		
		@Test
		public void testValidShortWithMessage() {
			
			e.expect(UnmetExpectationException.class);
			e.expectMessage("Custom: Expected 32768 to be a valid short value.");
			expect(Short.MAX_VALUE + 1).validShort("Custom");
		}
	}
}

File: core/src/test/scala/chaij/IntExpectationTests.scala
package chaij

import java.util.function.{ IntPredicate => IPred }

import org.scalatest._
import chaij.ChaiJ.expect

class IntExpectationTests extends MessageSpec {

  override type TestedType = Int
  override type ExpectationType = IntExpectation
  def expectationTypeName: String = "IntExpectation"
  override def expect(value: Int) = ChaiJ.expect(value)
  override def expect(value: Int, msg: String) = ChaiJ.expect(value, msg)
  def constructDummy(msg: String): (String, ExpectationType) = ("1", expect(1, msg))

  behavior of "Long expectations"

  registerGoodBadTests[Int](
    (42, _ equal _,
     ("make %d be equal to %d", 42 :: Nil),
     ("make %d not be equal to %d", 43 :: Nil)),
    (42, _ above _,
     ("make %d be above %d", 41 :: Nil),
     ("make %d not be above %d", 42 :: Nil)),
    (42, _ least _,
     ("make %d be at least %d", 42 :: Nil),
     ("make %d not be at least %d", 43 :: Nil)),
    (42, _ below _,
     ("make %d be below %d", 43 :: Nil),
     ("make %d not be below %d", 42 :: Nil)),
    (42, _ most _,
     ("make %d be at most %d", 42 :: Nil),
     ("make %d not be at most %d", 41 :: Nil)),
  )

  registerGoodBadTests[(Int, Int)](
    (42, (exp, tup) => exp.within(tup._1, tup._2),
     ("make %d be within %s", (42, 42) :: (41, 43) :: Nil),
     ("make %d not be within %s", (41, 41) :: (43, 43) :: Nil)),
  )

  registerGoodBadTests[IPred](
    (42, _ `match` _,
     ("make %d match _ == 42", ((_ == 42): IPred) :: Nil),
     ("make %d not match _ == 41",  ((_ == 41): IPred) :: Nil)),
    (42, _ satisfy _,
     ("make %d satisfy _ == 42", ((_ == 42): IPred) :: Nil),
     ("make %d not satisfy _ == 41",  ((_ == 41): IPred) :: Nil)),
  )

  they should "make 42 be close to 42 within 0" in {
    expect(42).closeTo(42, 0)
  }
  they should "make 42 be close to 41 and 43 within 1 each" in {
    expect(42).closeTo(41, 1).closeTo(43, 1)
  }
  they should "make 42 not be close to 41 within 0" in {
    an [UnmetExpectationException] should be thrownBy expect(42).closeTo(41, 0)
  }

  they should "make 42 be one of (41, 42, 43)" in {
    expect(42).oneOf(41, 42, 43)
  }
  they should "make 42 not be one of (41, 43)" in {
    an [UnmetExpectationException] should be thrownBy expect(42).oneOf(41, 43)
  }

  registerGoodBadNullaryTests(
    (_.validShort(),
      ("make %s be a valid Short",
        ("Short.MaxValue", Short.MaxValue.toInt) :: Nil),
      ("make %s not be a valid Short",
        ("Short.MaxValue + 1", Short.MaxValue.toInt + 1) ::
        ("Short.MinValue - 1", Short.MinValue.toInt - 1) :: Nil)),
    (_.validByte(),
      ("make %s be a valid Byte",
        ("Byte.MaxValue", Byte.MaxValue.toInt) :: Nil),
      ("make %s not be a valid Byte",
        ("Byte.MaxValue + 1", Byte.MaxValue.toInt + 1) ::
        ("Byte.MinValue - 1", Byte.MinValue.toInt - 1) :: Nil)),
  )

  "The int expectation messages" should "use the standard message" in {
    val exception = the [UnmetExpectationException] thrownBy {
      expect(1, "Special").equal(2)
    }
    exception.getMessage should startWith ("Special: ")
  }

  they should behave like registerMessageTests(
    (42, "equal", "Expected 42 to equal 43.",
      _.equal(43), _.equal(43, _)),
    (42, "above", "Expected 42 to be above 43.",
      _.above(43), _.above(43, _)),
    (42, "least", "Expected 42 to be at least 43.",
      _.least(43), _.least(43, _)),
    (42, "below", "Expected 42 to be below 41.",
      _.below(41), _.below(41, _)),
    (42, "most", "Expected 42 to be at most 41.",
      _.most(41), _.most(41, _)),
    (42, "within", "Expected 42 to be within 43 and 43."
      , _.within(43, 43), _.within(43, 43, _)),
    (42, "match", "Expected 42 to match a custom predicate.",
      _.`match`(_ => false), _.`match`(_ => false, _)),
    (42, "satisfy", "Expected 42 to satisfy a custom predicate.",
      _.satisfy(_ => false), _.satisfy(_ => false, _)),
    (42, "closeTo", "Expected 42 to be close to 43 with a delta of 0.",
      _.closeTo(43, 0), _.closeTo(43, 0, _)),
    (42, "oneOf", "Expected 42 to be one of [41, 43].",
      _.oneOf(41, 43), _.oneOf(_, 41, 43)),
    (Byte.MaxValue + 1, "validByte", "Expected 128 to be a valid byte value.",
      _.validByte(), _.validByte(_)),
    (Short.MaxValue + 1, "validShort", "Expected 32768 to be a valid short value.",
      _.validShort(), _.validShort(_)),
  )
}
